csharp/ql/src/semmle/code/csharp/dispatch/Dispatch.qll

```diff
@@ -249,15 +291,245 @@ private module Internal {
 
   abstract private class DispatchMethodOrAccessorCall extends DispatchCallImpl {
     pragma[nomagic]
-    predicate hasQualifierTypeInherited(SourceDeclarationType t) {
-      t = getAPossibleType(this.getQualifier(), _).getSourceDeclaration()
-    }
+    predicate hasQualifierTypeInherited(Type t) { t = getAPossibleType(this.getQualifier(), _) }
 
     pragma[nomagic]
     predicate hasQualifierTypeOverridden(ValueOrRefType t, OverridableCallable c) {
       hasQualifierTypeOverridden0(t, this) and
       hasCallable(any(OverridableCallable oc | hasQualifierTypeOverridden1(oc, this)), t, c)
     }
+
+    /**
+     * Holds if a call context may limit the set of viable source declaration
+     * run-time targets of this call.
+     *
+     * This is the case if the qualifier is either a `this` access or a parameter
+     * access, as the corresponding qualifier/argument in the call context may
+     * have a more precise type.
+     */
+    predicate mayBenefitFromCallContext(Callable c, int i) {
+      1 < strictcount(this.getADynamicTarget().getSourceDeclaration()) and
+      c = this.getCall().getEnclosingCallable().getSourceDeclaration() and
+      (
+        exists(AssignableDefinitions::ImplicitParameterDefinition pdef, Parameter p |
+          this.getQualifier() = BaseSsa::getARead(pdef, p) and
+          p.getPosition() = i and
+          c.getAParameter() = p and
+          not p.isParams()
+        )
+        or
+        i = -1 and
+        this.getQualifier() instanceof ThisAccess
```

Comment:
@aschackmull: You may want to handle the implicit `this` parameter in [the Java implementation](https://github.com/github/codeql/blob/master/java/ql/src/semmle/code/java/dataflow/internal/DataFlowDispatch.qll#L12) as well (unless `this` is already treated as a normal parameter).